### PR TITLE
create remote_ip_blocks in Source

### PR DESCRIPTION
**Pull request by kylebevans**
By adding remote_ip_blocks and not_remote_ip_blocks in Source,
an AuthorizationPolicy can trigger actions based on the original
client IP address gleaned from the X-Forwarded-For header or the
proxy protocol.

Reference: https://github.com/istio/istio/issues/22341

### CODE REVIEW

File: security/v1beta1/authorization_policy.proto [hunk 1]
```diff
@@ -503,6 +503,16 @@ message Source {
 
   // Optional. A list of negative match of IP blocks.
   repeated string not_ip_blocks = 8;
+
+  // Optional. A list of IP blocks, which matches to the "remote.ip" attribute.
+  // Populated from X-Forwarded-For or proxy protocol.
+  // Single IP (e.g. "1.2.3.4") and CIDR (e.g. "1.2.3.0/24") are supported.
+  //
+  // If not set, any IP is allowed.
+  repeated string remote_ip_blocks = 9;
```
Comment (yangminzhu) [inline security/v1beta1/authorization_policy.proto:512]: can you also update the comments of `source_ip` to make it clear that it is for the direct peer (not from X-Forwarded-For header)? thanks.

**Review by yangminzhu — COMMENTED** [review]
thanks, just a small comment.

**Comment by smawson** [comment]
Is this using anything in XFF or just the envoy trusted client ip? If trusted client maybe we can refer to that somewhere? We want to help users set this up in a way that is not spoofable.

File: security/v1beta1/authorization_policy.proto (hunk 1 above)
Comment (kylebevans) [reply security/v1beta1/authorization_policy.proto:513]: Yes, I have updated it in the latest commit.

**Comment by kylebevans** [comment]
@smawson Yes, it uses the envoy trusted client IP based on the numTrustedProxies setting in the meshConfig.  I am planning on updating the following docs to try to help:

https://istio.io/latest/docs/tasks/security/authorization/authz-ingress/
https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/
https://istio.io/latest/docs/reference/config/security/authorization-policy/

Are there other docs that should be updated?

**Comment by kylebevans** [comment]
@smawson I just realized the reference docs are generated from these files.  I updated the comments to reference numTrustedProxies and the article on Configuring Gateway Network Topology.

**Comment by smawson** [comment]
Cool, looks good to me, I'll wait for Yangmin to approve before I give the stamp.

File: security/v1beta1/authorization_policy.proto [hunk 2]
```diff
@@ -496,13 +496,27 @@ message Source {
...
   repeated string ip_blocks = 4;
 
   // Optional. A list of negative match of IP blocks.
   repeated string not_ip_blocks = 8;
+
+  // Optional. A list of IP blocks, which matches to the "remote.ip" attribute.
+  // Populated from X-Forwarded-For header or proxy protocol.
+  // To make use of this field, you must configure the numTrustedProxies field of the gatewayTopology under the meshConfig 
+  // when you install Istio or using an annotation on the ingress gateway.  See the documentation here:
+  // [Configuring Gateway Network Topology](https://istio.io/latest/docs/reference/config/security/authorization-policy/).
```
Comment (yangminzhu) [inline security/v1beta1/authorization_policy.proto:512]: The link should be https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/?

**Review by yangminzhu — COMMENTED** [review]
thanks, LGTM, just one small comment.

File: security/v1beta1/authorization_policy.proto (hunk 2 above)
Comment (kylebevans) [reply security/v1beta1/authorization_policy.proto:512]: whoops, yes, good catch.

**Review by yangminzhu — APPROVED** [review]
thanks!

**Review by smawson — APPROVED** [review]

**Pull request merged**